File: Lego_test.py
import random
from enum import IntEnum
from abc import ABC, abstractmethod

'''
-------------------------- IMPORTANT NOTES FOR CANDIDATES -------------------------
# 1) You are not expected to have to read or understand all of this code
# 2) You may use any existing language features, libraries, etc.
# 3) You may take time-saving shortcuts - but try to verbally indicate the "proper" way to do it
# 4) You may ask for help with anything - syntax, available library methods, where in the code X happens, etc.
# 5) You may compile & run, add print statements, and change things as much as you'd like.
------------------------------------------------------------------------------------
'''

'''
# The three moves you can throw in Rock, Paper, Scissors.
# Rock beats scissors.
# Scissors beats paper.
# Paper beats rock.
'''
class RPSMove(IntEnum):
    ROCK = 0
    SCISSORS = 1
    PAPER = 2

def handIsBetter(A, B):
    '''Return true iff move A beats move B.'''
    return (A+1)%3 == B


class Bot(ABC):
    '''A bot interface for playing Rock, Paper, Scissors.'''
    @abstractmethod
    def __init__(self):
        pass
    
    @abstractmethod    
    def makeMove(self):
        pass
    
    @abstractmethod
    def name(self):
        pass
    
    def onNewOpponent(self):
        pass
    
    def onHandPlayed(self, opponentsMove):
        pass

    
class randomBot(Bot):
    '''A bot that randomly chooses what to play each hand.'''
    def __init__(self):
        pass
    
    def name(self):
        return 'randomBot'
    
    def makeMove(self):
        selection = random.randint(0, len(RPSMove) - 1)
        action = RPSMove(selection)
        return action

class randomBot_2(Bot):
    '''A bot that randomly chooses what to play each hand.'''
    #20% rock, 30% paper, 50% scissors
    def __init__(self):
        pass
    
    def name(self):
        return 'randomBot_2'
    
    def makeMove(self):
        random_num = random.randint(0, 9)
        if random_num < 2:
            action = RPSMove(0)
        elif 2 <= random_num < 5:
            action = RPSMove(2)
        else:
            action = RPSMove(1)
        # selection = random.randint(0, len(RPSMove) - 1)

        # action = RPSMove(selection)
        return action


class oneMoveBot(Bot):
    '''A bot that always picks the same move every hand.'''
    def __init__(self, move):
        self.move = move
    
    def name(self):
        return 'oneMoveBot'
    
    def makeMove(self):
        action = RPSMove(self.move)
        return action

    
class basicStatsBot(Bot):
    '''
    A bot that tracks which hand their opponent has played the most, and plays the hand that will beat that.
    e.g. if their opponent has played more rock than anything else, it will play paper.
    '''
    def __init__(self):
        pass
    
    def name(self):
        return 'basicStatsBot'
    
    def makeMove(self):
        # Find the most played move.
        mostPlayedMoveCount = 0
        mostPlayedMove = 0
        for k, v in self.playedCounts.items():
            if v > mostPlayedMoveCount:
                mostPlayedMove = k
                mostPlayedMoveCount = v
                        
        # Find the counter to the most played move.
        for i in range(3):
            if handIsBetter(i, mostPlayedMove):
                return RPSMove(i)
    
    def onNewOpponent(self):
        self.playedCounts = {0:0, 1:0, 2:0}
    
    def onHandPlayed(self, opponentsMove):
        self.playedCounts[opponentsMove] += 1

    
class TournamentResults():
    '''Stores and displays the outcome of a tournament - who won, and what happened.'''
    
    def __init__(self, bots):

        keys = [bots[i].name() for i in range(len(bots))]
        self.botWins = dict(zip(keys, [0]*len(bots)))
        self.botLosses = dict(zip(keys, [0]*len(bots)))
        self.botTie = dict(zip(keys, [0]*len(bots)))
    
    def onGamesPlayed(self, botA, botB, winsA, winsB, winsAB):
        '''
        Adds the results of a set of games between two bots to these results.
        winsA is the number of times botA beat botB.
        winsB is the number of times botB beat botA.
        '''
        
        self.botWins[botA.name()] += winsA
        self.botLosses[botA.name()] += winsB
        self.botWins[botB.name()] += winsB
        self.botLosses[botB.name()] += winsA
        self.botTie[botA.name()] += winsAB
        self.botTie[botB.name()] += winsAB
      
    def printToConsole(self):
        '''Prints a summary of the tournament results to the console.'''
        
        print("{:20s} | {:7s} | {:7s} | {:7s} | {:4s}".format("Name", "Wins", "Losses", "Tie", "Overall"))
        print("---------------------------------------------------------------------\n")
        for i in range(len(self.botWins)):
            botName = list(self.botWins.keys())[i]
            print("{:20s} | {:7d} | {:7d} | {:7d} | {:4d}".format(
                            botName, 
                            self.botWins[botName], 
                            self.botLosses[botName], 
                            self.botTie[botName],
                            self.botWins[botName]-self.botLosses[botName]
                            ))

                                               
class Tournament():
    '''Runs a rock paper scissors tournament between a set of bots.'''
    
    def __init__(self, bots):
        self.bots = bots
        
        self.TournamentObserver = TournamentResults(bots)
               
    def _playGames(self, playerA, playerB, play_times): 
        '''Plays a series of games between these two bots, returning how many times each bot won.'''
        
        wins = [0, 0, 0] # [A win, B win]
        playerA.onNewOpponent()
        playerB.onNewOpponent()
        print("---------")
        print(f"{playerA.name()} v.s. {playerB.name()}:")
        # range(1000)
        for i in range(play_times):
            moveA = playerA.makeMove()
            moveB = playerB.makeMove()

            if handIsBetter(moveA, moveB):
                wins[0] += 1
                print(f"{playerA.name()}: A wins")
            elif handIsBetter(moveB, moveA):
                wins[1] += 1
                print(f"{playerB.name()}: B wins")
            else:
                wins[2] += 1
                print(f"Tie")
            
            playerA.onHandPlayed(moveB)
            playerB.onHandPlayed(moveA)

        return wins
         
    def playTournament(self, play_times):
        '''Plays through the tournament, having each bot play against the others in a round-robin format.'''
        
        bot_size = len(self.bots)
        
        # Go through each unique pair of bots.
        for i in range(bot_size):
            for j in range(i+1,bot_size):                
                # Play the games.
                wins = self._playGames(self.bots[i], self.bots[j], play_times)
                # Send the results to any observers.
                self.TournamentObserver.onGamesPlayed(self.bots[i], self.bots[j], wins[0], wins[1], wins[2])
                
        self.TournamentObserver.printToConsole()
        
          
def main():
    play_times_2 = 10
    RandomBot = randomBot()
    OneMoveBot = oneMoveBot(RPSMove.SCISSORS)
    BasicStatsBot = basicStatsBot()
    RandomBot_2 = randomBot_2()

    # Setup our tournament.
    bots = [RandomBot, OneMoveBot, BasicStatsBot, RandomBot_2]
    # tourny = Tournament(bots)
    tourny_2 = Tournament(bots)

    # Play the tournament.
    # tourny.playTournament(play_times)
    tourny_2.playTournament(play_times_2)


    
if __name__ == "__main__":
    main()

File: Lego_test.cc
#include <iostream>
#include <stdlib.h>
#include <string>
#include <map>
#include <vector>

using namespace std;

/*********************************************************
 * IMPORTANT NOTES:
 * 1) You are not expected to have to read or understand all of this code
 * 2) You may use any existing language features, libraries, etc.
 * 3) You may take time-saving shortcuts - but try to verbally indicate the "proper" way to do it
 * 4) You may ask for help with anything - syntax, available library methods, where in the code X happens, etc.
 * 5) You may compile & run, add print statements, and change things as much as you'd like.
 */



/**
 * The three moves you can throw in Rock, Paper, Scissors.
 * Rock beats scissors.
 * Scissors beats paper.
 * Paper beats rock.
 */
enum RPSMove {
    ROCK,
    SCISSORS,
    PAPER
};

/**
 * Return true iff move A beats move B.
 */
static bool handIsBetter(RPSMove a, RPSMove b) {
    return ((a+1)%3) == b;
}


/**
 * A bot interface for playing Rock, Paper, Scissors.
 */
class Bot {
public:
    virtual ~Bot() {}

    /**
     * An identifier for what type of bot this is.
     */
    virtual string name() const = 0;

    /**
     * Returns this bot's next throw against the current opponent.
     */
    virtual RPSMove makeMove() = 0;

    /**
     * Invoked prior to a series of games with a new opponent.
     */
    virtual void onNewOpponent() {}

    /**
     * Invoked after each hand against an opponent, indicating what move the opponent made.
     */
    virtual void onHandPlayed(RPSMove opponentsMove) {}
};


/**
 * Class for receiving information about what happens during the tournament.
 */
class TournamentObserver {
public:
    virtual ~TournamentObserver() {}

    /**
     * Invoked after a series of games are played between two bots.
     */
    virtual void onGamesPlayed(const Bot& botA, const Bot& botB, int winsA, int winsB) = 0;
};

/**
 * Runs a rock paper scissors tournament between a set of bots.
 */
class Tournament {
public:
    Tournament(vector<Bot*> bots)
        : _bots(std::move(bots)) {
    }

    /**
     * Adds an observer to be notified when events happen in the tournament.
     */
    void addObserver(TournamentObserver* observer) {
        _observers.push_back(observer);
    }

    /**
     * Plays through the tournament, having each bot play against the others in a round-robin format.
     */
    void playTournament() {
        static int numGames = 1000;
        // Go through each unique pair of bots.
        for (size_t i = 0; i < _bots.size(); ++i) {
            for (size_t j = i + 1; j < _bots.size(); ++j) {

                // Play the games.
                pair<int, int> wins = playGames(*_bots[i], *_bots[j], numGames);

                // Send the results to any observers.
                for (auto* observer : _observers) {
                    observer->onGamesPlayed(*_bots[i], *_bots[j], wins.first, wins.second);
                }
            }
        }
    }

private:
    /**
     * Plays a series of games between these two bots, returning how many times each bot won.
     */
    pair<int, int> playGames(Bot& playerA, Bot& playerB, int numGames) {
        pair<int, int> wins;
        playerA.onNewOpponent();
        playerB.onNewOpponent();
        
        for (int i = 0; i < numGames; ++i) {
            RPSMove moveA = playerA.makeMove();
            RPSMove moveB = playerB.makeMove();
            if (handIsBetter(moveA, moveB)) {
                wins.first++;
            } else if (handIsBetter(moveB, moveA)) {
                wins.second++;
            }

            playerA.onHandPlayed(moveB);
            playerB.onHandPlayed(moveA);
        }

        return wins;
    }

    vector<Bot*> _bots;
    vector<TournamentObserver*> _observers;
};

/**
 * Stores and displays the outcome of a tournament - who won, and what happened.
 */
class TournamentResults : public TournamentObserver {
public:

    /**
     * Adds the results of a set of games between two bots to these results.
     * winsA is the number of times botA beat botB.
     * winsB is the number of times botB beat botA.
     */
    void onGamesPlayed(const Bot& botA, const Bot& botB, int winsA, int winsB) override {
        _botWins[botA.name()] += winsA;
        _botLosses[botA.name()] += winsB;
        
        _botWins[botB.name()] += winsB;
        _botLosses[botB.name()] += winsA;
    }

    /**
     * Prints a summary of the tournament results to the console.
     */
    void printToConsole() {
        printf("%20s | %7s | %7s | %4s\n", "Name", "Wins", "Losses", "Overall");
        printf("--------------------------------------------------\n");
        for (const auto& entry : _botWins) {
            auto botName = entry.first;

            printf("%20s | %7d | %7d | %4d\n", botName.c_str(), _botWins[botName], _botLosses[botName], _botWins[botName] - _botLosses[botName]);
        }
    }

private:
    // Stores total wins across the tournament for each bot, by name.
    map<string, int> _botWins;

    // Stores total losses across the tournament for each bot, by name.
    map<string, int> _botLosses;
};

/**
 * A bot that randomly chooses what to play each hand.
 */
class RandomBot : public Bot {
public:
    string name() const override {
        return "RandomBot";
    }
    RPSMove makeMove() override {
        return (RPSMove)(rand() % 3);
    }
};

/**
 * A bot that always picks the same move every hand.
 */
class OneMoveBot : public Bot {
public:
    OneMoveBot(RPSMove move) : _move(move) {
    }

    string name() const override {
        return "OneMoveBot";
    }

    RPSMove makeMove() override {
        return _move;
    }

private:
    RPSMove _move;
};


/**
 * A bot that tracks which hand their opponent has played the most, and plays
 * the hand that will beat that.
 *
 * e.g. if their opponent has played more rock than anything else, it will play paper.
 */
class BasicStatsBot : public Bot {
public:
    string name() const override {
        return "BasicStatsBot";
    }

    RPSMove makeMove() override {
        // Find the most played move.
        RPSMove mostPlayedMove = ROCK;
        int mostPlayedMoveCount = 0;
        for (const auto& cnt : _playedCounts) {
            if (cnt.second > mostPlayedMoveCount) {
                mostPlayedMove = cnt.first;
                mostPlayedMoveCount = cnt.second;
            }
        }

        // Find the counter to the most played move.
        for (int i = 0; i < 3; ++i) {
            if (handIsBetter((RPSMove)i, mostPlayedMove)) {
                // Found the counter, return it.
                return (RPSMove)i;
            }
        }

        throw "This should be impossible!";
    }

    void onNewOpponent() override {
        _playedCounts.clear();
    }

    void onHandPlayed(RPSMove opponentsMove) override {
        _playedCounts[opponentsMove]++;
    }

private:
    map<RPSMove, int> _playedCounts;
};

int main() {
    // Initialize our list of bots.
    RandomBot randomBot;
    OneMoveBot scissorsBot(SCISSORS);
    BasicStatsBot basicStatsBot;

    // Setup our tournament.
    {
        Tournament tourny({&randomBot, &scissorsBot, &basicStatsBot});
        
        // Keep track of results.
        TournamentResults results;
        tourny.addObserver(&results);
        
        // Play the tournament.
        tourny.playTournament();
        
        // Print out the results.
        results.printToConsole();
    }
}
